src/dimension.c

```diff
@@ -1558,6 +1550,38 @@ ts_dimension_add(PG_FUNCTION_ARGS)
 
 		/* Check that partitioning is sane */
 		ts_hypertable_check_partitioning(info.ht, dimension_id);
+
+		/*
+		 * If the hypertable has chunks, to make it compatible
+		 * we add artificial dimension slice which will cover -inf / inf
+		 * range.
+		 *
+		 * Newly created chunks will have a proper slice range according to
+		 * the created dimension and its partitioning.
+		 */
+		if (ts_hypertable_has_chunks(info.table_relid, AccessShareLock))
+		{
+			ListCell *lc;
+			List *chunk_id_list = ts_chunk_get_chunk_ids_by_hypertable_id(info.ht->fd.id);
+
+			DimensionSlice *slice;
+			slice = ts_dimension_slice_create(dimension_id,
+											  DIMENSION_SLICE_MINVALUE,
+											  DIMENSION_SLICE_MAXVALUE);
+			ts_dimension_slice_insert_multi(&slice, 1);
+
+			foreach (lc, chunk_id_list)
+			{
+				int chunk_id = lfirst_int(lc);
+				Chunk *chunk = ts_chunk_get_by_id(chunk_id, true);
+				ChunkConstraint *cc = ts_chunk_constraints_add(chunk->constraints,
```

Comment:
What's the locking semantics here? Do we need a lock on the chunk since we're making modifications?

Often the semantics are:

Check for chunk -> get lock -> check again that chunk exists after locking

Reply:
Would be good to have an isolation test checking for locking/concurrency issues, for example:

* Running concurrent add_dimension() on the same table in two different session
* Running add_dimension() concurrently with creating new chunk (obviously the new chunk should also have the dimension)

This can perhaps be achieved by adding a waitlock in this code, after retrieving the "Chunk *" but before adding the constraints. Then running an insert to create a new chunk (and/or, calling explicit chunk creation function used by copy/move chunks)

Reply:
Good point, will do